netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -107,19 +109,29 @@ public static ProtocolNegotiator serverPlaintext() {
   /**
    * Create a server TLS handler for HTTP/2 capable of using ALPN/NPN.
    */
-  public static ProtocolNegotiator serverTls(final SslContext sslContext) {
+  public static ProtocolNegotiator serverTls(final SslContext sslContext,
+      final ObjectPool<? extends Executor> executorPool) {
     Preconditions.checkNotNull(sslContext, "sslContext");
+    final Executor executor;
+    if (executorPool != null) {
+      executor = executorPool.getObject();
+    } else {
+      executor = null;
+    }
     return new ProtocolNegotiator() {
       @Override
       public ChannelHandler newHandler(GrpcHttp2ConnectionHandler handler) {
         ChannelHandler gnh = new GrpcNegotiationHandler(handler);
-        ChannelHandler sth = new ServerTlsHandler(gnh, sslContext);
+        ChannelHandler sth = new ServerTlsHandler(gnh, sslContext, executor);
```

Comment:
The handlers here can out-live the `ProtocolNegotiator`. So they need to get their own reference to `executor` from `executorPool`. I suggest passing in `executorPool` here and calling `getObject()` within the constructor.

You would still keep the rest of this protocol negotiator the same; you'd still call `getObject()` within `serverTls()` and `returnObject()` in `close()`. Keeping the extra (unused) reference here forces the executor to stay alive, which prevents it from being re-created for every connection.

Reply:
may be it is obvious but I thought I should mention this: with this additional call to `getObject()` in the `ServerTlsHandler` constructor you will also need a `returnObject()` call within `ServerTlsHandler` somewhere - you may need to implement `handlerRemoved` just for this.

I didn't completely understand the comment "Keeping the extra (unused) reference here forces the executor to stay alive, which prevents it from being re-created for every connection."

Reply:
We expect the Pool here to generally be a SharedResourcePool. It releases a resource when it is no longer used, so if we don't hold a reference to the executor it may be created for a connection, shutdown after the connection handshake, created again for the next connection, shutdown after the connection handshake, etc.

Reply:
Thanks a lot for the valuable discussion here!
I have updated as suggested but am not 100% confident that I understand correctly.
Previously, I thought the Pool would be a collection of objects, and every time a getObject() call will return a completely new Object. This requires us to call returnObject() for every object got from getObject().
It looks like for a SharedResourcePool, the new object will only be created when the old one is released(by GC, etc). The number of calling getObject() and returnObject() doesn't necessarily to be the same.
Please correct me if I am wrong. Thanks again!